x/distribution/src/abci_handler.rs

```diff
@@ -44,6 +73,128 @@ impl<
+    ) -> Result<(), AppError> {
+        match msg {
+            Message::WithdrawRewards(msg) => self
+                .keeper
+                .withdraw_delegator_reward_and_commission(ctx, msg),
+            Message::SetWithdrawAddr(msg) => self.keeper.set_withdraw_address(ctx, msg),
+            Message::FundCommunityPool(msg) => self.keeper.fund_community_pool_cmd(ctx, msg),
+        }
+    }
+
+    pub fn query<DB: Database + Send + Sync>(
+        &self,
+        ctx: &QueryContext<DB, SK>,
+        query: RequestQuery,
+    ) -> Result<prost::bytes::Bytes, AppError> {
+        match query.path.as_str() {
+            "/cosmos.distribution.v1beta1.Query/ValidatorOutstandingRewards" => {
+                let req = QueryValidatorOutstandingRewardsRequest::decode(query.data)
+                    .map_err(|e| CoreError::DecodeProtobuf(e.to_string()))?;
+
+                Ok(self
+                    .keeper
+                    .query_validator_outstanding_rewards(ctx, req)
+                    .encode_vec()
+                    .into())
+            }
+            "/cosmos.distribution.v1beta1.Query/ValidatorCommission" => {
+                let req = QueryValidatorCommissionRequest::decode(query.data)
+                    .map_err(|e| CoreError::DecodeProtobuf(e.to_string()))?;
+
+                Ok(self
+                    .keeper
+                    .query_validator_commission(ctx, req)
+                    .encode_vec()
+                    .into())
+            }
+            "/cosmos.distribution.v1beta1.Query/ValidatorSlashes" => {
+                let req = QueryValidatorSlashesRequest::decode(query.data)
+                    .map_err(|e| CoreError::DecodeProtobuf(e.to_string()))?;
+
+                Ok(self
+                    .keeper
+                    .query_validator_slashes(ctx, req)
+                    .encode_vec()
+                    .into())
+            }
+            "/cosmos.distribution.v1beta1.Query/DelegationRewards" => {
+                let req = QueryDelegationRewardsRequest::decode(query.data)
+                    .map_err(|e| CoreError::DecodeProtobuf(e.to_string()))?;
+
+                Ok(self
+                    .keeper
+                    .query_delegation_rewards(ctx, req)?
+                    .encode_vec()
+                    .into())
+            }
+            "/cosmos.distribution.v1beta1.Query/DelegatorValidators" => {
+                let req = QueryWithdrawAllRewardsRequest::decode(query.data)
+                    .map_err(|e| CoreError::DecodeProtobuf(e.to_string()))?;
+
+                Ok(self
+                    .keeper
+                    .query_delegator_validators(ctx, req)
+                    .encode_vec()
+                    .into())
+            }
+            "/cosmos.distribution.v1beta1.Query/CommunityPool" => {
+                let req = QueryCommunityPoolRequest::decode(query.data)
+                    .map_err(|e| CoreError::DecodeProtobuf(e.to_string()))?;
```

Comment:
Note: if you merge upstream line `.map_err(|e| CoreError::DecodeProtobuf(e.to_string()))?;` could be omitted.